vernac/declaremods.ml

```diff
@@ -822,57 +1084,114 @@ let declare_module id args res mexpr_o fs =
   let () = check_subtypes mp subs in
 
   let sobjs = subst_sobjs (map_mp mp0 mp resolver) sobjs in
-  add_leaf (ModuleObject (id,sobjs));
+  InterpVisitor.add_leaf (ModuleObject (id,sobjs));
   mp
 
 end
 
+end
+
 (** {6 Module types : start, end, declare} *)
 
 module RawModTypeOps = struct
 
+module Synterp = struct
+
+let start_modtype_core id cur_mp args mtys fs =
+  let mp = ModPath.MPdot(cur_mp, Label.of_id id) in
+  let args = RawModOps.Synterp.intern_args args in
+  let mbids = List.flatten @@ List.map (fun (mbidl,_) -> mbidl) args in
+  let sub_mty_l = RawModOps.Synterp.build_subtypes mtys in
+  mp, mbids, args, sub_mty_l
+
+let start_modtype id args mtys fs =
+  let mp, mbids, args, sub_mty_l = start_modtype_core id !openmod_syntax_info.cur_mp args mtys fs in
+  openmod_syntax_info := { cur_mp = mp; cur_typ = None; cur_mbids = mbids };
+  let prefix = Lib.Synterp.start_modtype id mp fs in
+  Nametab.(push_dir (Until 1) (prefix.obj_dir) (GlobDirRef.DirOpenModtype prefix.obj_mp));
+  mp, args, sub_mty_l
+
+let end_modtype_core id mbids objects fs =
+  let {Lib.substobjs = substitute; keepobjs = _; anticipateobjs = special; } = objects in
+  Summary.unfreeze_summaries ~partial:true fs;
+  let modtypeobjs = (mbids, Objs substitute) in
+  (special@[ModuleTypeObject (id,modtypeobjs)])
+
+let end_modtype () =
+  let oldprefix,fs,objects = Lib.Synterp.end_modtype () in
+  let olddp, id = split_dirpath oldprefix.Nametab.obj_dir in
+  let objects = end_modtype_core id !openmod_syntax_info.cur_mbids objects fs in
+  SynterpVisitor.add_leaves objects;
+  !openmod_syntax_info.cur_mp
+
+(* FIXME use open_modtype_core *)
+let declare_modtype id args mtys (mty,ann) fs =
+  let inl = inl2intopt ann in
+  (* We simulate the beginning of an interactive module,
+     then we adds the module parameters to the global env. *)
+  let mp = ModPath.MPdot(!openmod_syntax_info.cur_mp, Label.of_id id) in
+  let args = RawModOps.Synterp.intern_args args in
+  let mbids = List.flatten @@ List.map fst args in
+  let mte, base, kind = Modintern.intern_module_ast Modintern.ModType mty in
+  let entry = mbids, mte in
+  let sub_mty_l = RawModOps.Synterp.build_subtypes mtys in
+  let sobjs = RawModOps.Synterp.get_functor_sobjs false inl entry in
+  let subst = map_mp (get_module_path (snd entry)) mp empty_delta_resolver in
+  let sobjs = subst_sobjs subst sobjs in
+
+  (* Undo the simulated interactive building of the module type
+     and declare the module type as a whole *)
+  Summary.unfreeze_summaries ~partial:true fs; (* TODO check this *)
```

Comment:
Is this TODO intended to be solved before merging?

Reply:
No, it was actually the same as @gares ' question. I removed the TODO.